src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -58,57 +58,169 @@ internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
-        List<SemanticRange>? csharpSemanticRanges = null;
-
+        List<SemanticRange>? razorSemanticRanges = null;
+        PooledObject<List<SemanticRange>>? csharpSemanticRanges = null;
+        
         try
         {
-            csharpSemanticRanges = await GetCSharpSemanticRangesAsync(codeDocument, textDocumentIdentifier, range, razorSemanticTokensLegend, documentContext.Version, correlationId, cancellationToken).ConfigureAwait(false);
+            var csharpSemanticRangesTask = GetCSharpSemanticRangesAsync(
+                codeDocument, textDocumentIdentifier, range, razorSemanticTokensLegend, documentContext.Version, correlationId, cancellationToken);
+
+            razorSemanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument, range, razorSemanticTokensLegend, _razorLSPOptionsMonitor.CurrentValue.ColorBackground);
+            csharpSemanticRanges = await csharpSemanticRangesTask.ConfigureAwait(false);
         }
         catch (Exception ex)
         {
             _logger.LogError(ex, "Error thrown while retrieving CSharp semantic range.");
         }
 
-        var combinedSemanticRanges = CombineSemanticRanges(razorSemanticRanges, csharpSemanticRanges);
+        if (cancellationToken.IsCancellationRequested)
+        {
+            if (csharpSemanticRanges.HasValue)
+            {
+                csharpSemanticRanges.Value.Dispose();
+            }
+
+            return null;
+        }
+
+        using var combinedSemanticRanges = CombineSemanticRanges(razorSemanticRanges, csharpSemanticRanges is not null ? csharpSemanticRanges.Value.Object : null);
 
         // We return null when we have an incomplete view of the document.
         // Likely CSharp ahead of us in terms of document versions.
         // We return null (which to the LSP is a no-op) to prevent flashing of CSharp elements.
-        if (combinedSemanticRanges is null)
+        if (combinedSemanticRanges is null || !combinedSemanticRanges.HasValue)
         {
             _logger.LogWarning("Incomplete view of document. C# may be ahead of us in document versions.");
+
+            if (csharpSemanticRanges.HasValue)
+            {
+                csharpSemanticRanges.Value.Dispose();
+            }
+
             return null;
         }
 
-        var data = ConvertSemanticRangesToSemanticTokensData(combinedSemanticRanges, codeDocument);
+        var data = ConvertSemanticRangesToSemanticTokensData(combinedSemanticRanges.Value.Object, codeDocument);
         var tokens = new SemanticTokens { Data = data };
 
+        if (csharpSemanticRanges.HasValue)
+        {
+            csharpSemanticRanges.Value.Dispose();
+        }
+
         return tokens;
     }
 
-    private static List<SemanticRange>? CombineSemanticRanges(List<SemanticRange>? ranges1, List<SemanticRange>? ranges2)
+    private static PooledObject<List<SemanticRange>>? CombineSemanticRanges(List<SemanticRange>? razorRanges, List<SemanticRange>? csharpRanges)
```

Comment:
I don't see the value of returning the pool here, versus just returning a nullable list. It makes the calling code much more confusing, having to deal with disposal in various paths, and doesn't seem to save anything.